microsoft/CBL-Mariner
# Add LiveOS ISO creation support to MIC.

###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*

---

###### Summary 

Given a full disk image, the Mariner Image Customizer (MIC) can generate a
LiveOS ISO image when the `--output-image-format` is set to `iso`.

The LiveOS ISO image is a bootable image that boots into the rootfs of the
input full disk image. This eliminates the requirement to install that image
to persistent storage before booting it.

All customizations can still be made to the input full disk image rootfs as
usual using MIC, and will naturally carry over into the LiveOS ISO. This
includes customizations for kernel modules, dracut, and other early boot-time
actions.

While converting the input full disk image into a LiveOS ISO involves copying
almost all the artifacts unchanged - some artifacts are changed as follows:
- `grub.cfg` is replaced with a stock one that serves the LiveOS boot flow.
  This can be important to account for if you rely on certain grub
  configurations in your current grub.cfg.
- `/etc/fstab` is dropped from the rootfs as it typically conflicts with the
  overlay setup required by the LiveOS.
- `initrd.img` is regenerated to serve the LiveOS boot flow. This should have
  no impact as long as the included rootfs is where the original `initrd.img`
  was generated using `dracut`. The use of such rootfs guarantees that the same
  dracut configuration that got used before will be re-used again when we are
  re-generating the `initrd.img` and hence the original behavior is retained.

The current implementation for the LiveOS ISO does not support persistent
storage. This means that the machine loses all its state on reboot or
shutdown.

###### Change Log  



- Add LiveOS ISO creation support to MIC.

###### Does this affect the toolchain?  



**NO**

###### Associated issues  


- [6167 Update MIC to customize an initrd image](https://dev.azure.com/mariner-org/ECF/_workitems/edit/6167)

###### Links to CVEs  
- n/a

###### Test Methodology

- Build the new dracut package.
- Build baremetal image.
- Create a MIC config to add a user name/password.
- Use MIC to customize the baremetal image using the config and specify the output to be iso.
- Boot the iso in a VM.
- Log-in and verify the overlays.
